File: Rivington.IG.Web/Rivington.IG.Domain.Models/OrderManagement/InspectionOrderWildfireAssessment/Roof.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace Rivington.IG.Domain.Models.OrderManagement
{
    public class InspectionOrderWildfireAssessmentRoof
    {
        [Parent]
        public InspectionOrderWildfireAssessment InspectionOrderWildfireAssessment { get; set; }
        [Key]
        public Guid Id { get; set; }
        

        public string PrimaryRoofCoveringId { get; set; }
        [ForeignKey(nameof(PrimaryRoofCoveringId))]
        public PrimaryRoofCovering PrimaryRoofCovering { get; set; }

        public string SecondaryRoofCoveringId { get; set; }
        [ForeignKey(nameof(SecondaryRoofCoveringId))]
        public SecondaryRoofCovering SecondaryRoofCovering { get; set; }

        public ICollection<InspectionOrderWildfireAssessmentRoofOtherRoofCoverings> OtherRoofCoverings { get; set; }

        public string RoofTypeId { get; set; }
        [ForeignKey(nameof(RoofTypeId))]
        public RoofType RoofType { get; set; }

        public string RoofShapeId { get; set; }
        [ForeignKey(nameof(RoofShapeId))]
        public RoofShape RoofShape { get; set; }

        public string RoofPitchId { get; set; }
        [ForeignKey(nameof(RoofPitchId))]
        public RoofPitch RoofPitch { get; set; }

        public ICollection<InspectionOrderWildfireAssessmentRoofRoofConcernDetails> RoofConcernDetails { get; set; }
        public bool CombustibleMaterialsOnRoof { get; set; }
        public string CombustibleMaterialsonRoofComment { get; set; }
        public ICollection<InspectionOrderWildfireAssessmentRoofRoofVentings> RoofVentings { get; set; }
        public bool VentingOpeningAllowEmberEntry { get; set; }
        public string VentingMeshCoveringSizeComment { get; set; }
        public bool Gutter { get; set; }
        public ICollection<InspectionOrderWildfireAssessmentRoofGutterTypes> GutterTypes { get; set; }
        public string GutterComment { get; set; }

        public string EavesTypeId { get; set; }
        [ForeignKey(nameof(EavesTypeId))]
        public EavesType EavesType { get; set; }

        public bool EavesAllowEmberEntry { get; set; }
        public string EavesComment { get; set; }

        public static void BuildModel(ModelBuilder modelBuilder)
        {
            // IoWa = Inspection Order Wildfire Assessment
            const string entTypeName = "IoWaRoof";
            modelBuilder.Entity<InspectionOrderWildfireAssessment>()
                .HasOne(a => a.Roof).WithOne(b => b.InspectionOrderWildfireAssessment)
                .HasForeignKey<InspectionOrderWildfireAssessmentRoof>()
                .HasConstraintName($"FK_{nameof(InspectionOrderWildfireAssessment)}_{entTypeName}")
                .OnDelete(DeleteBehavior.Cascade);

            var entBuilder = modelBuilder.Entity<InspectionOrderWildfireAssessmentRoof>();
            entBuilder.HasOne(a => a.PrimaryRoofCovering).WithMany().HasConstraintName($"FK_{entTypeName}_{nameof(PrimaryRoofCovering)}");
            entBuilder.HasOne(a => a.SecondaryRoofCovering).WithMany().HasConstraintName($"FK_{entTypeName}_{nameof(SecondaryRoofCovering)}");
            
            InspectionOrderWildfireAssessmentRoofOtherRoofCoverings.BuildModel(modelBuilder);

            entBuilder.HasOne(a => a.RoofType).WithMany().HasConstraintName($"FK_{entTypeName}_{nameof(RoofType)}");
            entBuilder.HasOne(a => a.RoofShape).WithMany().HasConstraintName($"FK_{entTypeName}_{nameof(RoofShape)}");
            entBuilder.HasOne(a => a.RoofPitch).WithMany().HasConstraintName($"FK_{entTypeName}_{nameof(RoofPitch)}");

            InspectionOrderWildfireAssessmentRoofRoofConcernDetails.BuildModel(modelBuilder);

            InspectionOrderWildfireAssessmentRoofRoofVentings.BuildModel(modelBuilder);

            InspectionOrderWildfireAssessmentRoofGutterTypes.BuildModel(modelBuilder);

            entBuilder.HasOne(a => a.EavesType).WithMany().HasConstraintName($"FK_{entTypeName}_{nameof(EavesType)}");
        }
    }
}

File: Rivington.IG.Web/Rivington.IG.API/src/app/modules/shared/models/ordermanagement/inspection-order/checklist/wildfire/roof.ts
import { PrimaryRoofCovering } from "./roof/primary-roof-covering";
import { InspectionOrderWildfireAssessmentRoofOtherRoofCoverings } from "./roof/InspectionOrderWildfireAssessmentRoofOtherRoofCoverings";
import { InspectionOrderWildfireAssessmentRoofRoofConcernDetails } from "./roof/InspectionOrderWildfireAssessmentRoofRoofConcernDetails";
import { InspectionOrderWildfireAssessmentRoofRoofVentings } from "./roof/InspectionOrderWildfireAssessmentRoofRoofVentings";
import { InspectionOrderWildfireAssessmentRoofGutterTypes } from "./roof/InspectionOrderWildfireAssessmentRoofGutterTypes";
import { EavesType } from "./roof/eaves-type";
import { SecondaryRoofCovering } from "./roof/secondary-roof-covering";
import { RoofType } from "./roof/roof-type";
import { InspectionOrderWildfireAssessment } from "../wildfire";
import { RoofShape } from "../property/home-detail/roof-shape";
import { RoofPitch } from "./roof/roof-pitch";

export class InspectionOrderWildfireAssessmentRoof {
    /*[Parent]*/
    inspectionOrderWildfireAssessment?: InspectionOrderWildfireAssessment;
    id?: string;
    primaryRoofCoveringId?: string;
    // [ForeignKey(nameof(PrimaryRoofCoveringId))]
    primaryRoofCovering?: PrimaryRoofCovering;
    secondaryRoofCoveringId?: string;
    // [ForeignKey(nameof(SecondaryRoofCoveringId))]
    secondaryRoofCovering?: SecondaryRoofCovering;
    otherRoofCoverings?: InspectionOrderWildfireAssessmentRoofOtherRoofCoverings[];
    roofShapeId?: string;
    // [ForeignKey(nameof(roofShapeId))]
    roofShape?: RoofShape;
    roofPitchId?: string;
    // [ForeignKey(nameof(roofPitchId))]
    roofPitch?: RoofPitch;
    roofConcernDetails?: InspectionOrderWildfireAssessmentRoofRoofConcernDetails[];
    combustibleMaterialsOnRoof?: boolean;
    combustibleMaterialsonRoofComment?: string;
    roofVentings?: InspectionOrderWildfireAssessmentRoofRoofVentings[];
    ventingOpeningAllowEmberEntry?: boolean;
    ventingMeshCoveringSizeComment?: string;
    gutter?: boolean;
    gutterTypes?: InspectionOrderWildfireAssessmentRoofGutterTypes[];
    gutterComment?: string;
    eavesTypeId?: string;
    // [ForeignKey(nameof(EavesTypeId))]
    eavesType?: EavesType;
    eavesAllowEmberEntry?: boolean;
    eavesComment?: string;
    
    public constructor(init?:Partial<InspectionOrderWildfireAssessmentRoof>) {
        Object.assign(this, init);
    }
}